Added timeout lock to readthedocs

jlashner opened the pull request:
Added a TimeoutLock page to the readthedocs. Closes #115.

BrianJKoopman commented on docs/developer/timeout_lock.rst:
```diff
@@ -0,0 +1,52 @@
+
+In OCS, operations are mainly asynchronous, however most agents have
+restrictions on which operations are allowed to be run simultaneously.
+For instance, when only one operation can query with a hardware device at a time
```
I think this could be reworded, how about:
"For instance, when an Agent interfaces with a hardware device over serial communication, two commands cannot be issued simultaneously."

BrianJKoopman commented on docs/developer/timeout_lock.rst:
```diff
@@ -0,0 +1,52 @@
+an acquisition process is in progress is dangerous.
+
+In these cases, it is required for agents to create a `lock` which restricts
+what operations can run simultaneously. For instance, in your agent it may
```
Another wording suggestion for the sentence that starts with "For instance..." -> "In our example, it is necessary for each operation that communicates over with the hardware device to acquire the lock."

BrianJKoopman commented on docs/developer/timeout_lock.rst:
```diff
@@ -0,0 +1,52 @@
+
+The ocs TimeoutLock is a lock that implements a few useful additions to the
+standard ``threading.lock``. Mainly:
+ -  It can be acquired the ``acquire_timeout`` context
```
There might be a word missing here?

BrianJKoopman commented on docs/developer/timeout_lock.rst:
```diff
@@ -0,0 +1,52 @@
+standard ``threading.lock``. Mainly:
+ -  It can be acquired the ``acquire_timeout`` context
+    manager, which will automatically release even if an error is raised.
+ -  A ``job`` string can be set on acquisition, so it is clear why an acquisition
```
Change after the comma to "describing the job obtaining the lock." This gives developers guidance for what they should call the job.

BrianJKoopman commented on docs/developer/timeout_lock.rst:
```diff
@@ -0,0 +1,52 @@
+    so operations can safely run during long-running processes
+ -  A ``timeout`` can be set to limit how long an operation will wait for the
+    lock before failing.
+
```
Add new section for Examples (at the '---' level.) Will provide room for when `YieldingLock` gets moved.

BrianJKoopman replied:
Ah, in Slack I was told the functionality was moved into TimeoutLock already. A more advanced example of using the lock to allow a long running process that requires the lock to yield to a task that needs to communicate intermittently would be very valuable. Much like is done in the 372 Agent now.

BrianJKoopman commented on docs/developer/timeout_lock.rst:
```diff
@@ -0,0 +1,52 @@
+----
+
+.. autoclass:: ocs.ocs_twisted.TimeoutLock
+    :members:
```
The API is autodoc'd already in the [API Reference](https://ocs.readthedocs.io/en/latest/api.html). This causes a bunch of warnings for duplicate autodocs. I've solved this in previous instances by putting `:noindex:` on the page reference, i.e. here. Searches will yield the API page in that case.

BrianJKoopman commented on docs/index.rst:
```diff
@@ -46,6 +46,7 @@ write OCS Agents or Clients.
     developer/feeds
     developer/agents
     developer/clients
+    developer/timeout_lock
```
I'm not sure this should get its own top level in the sidebar, I think it fits better under Agents. Though as I say that I realize the Agents page should probably be broken up a bit. I do think we should move this in there, whether now or later.

jlashner commented:
Hey Brian, thanks for the suggestions. I reworded the parts you mentioned. Let me know if you still don't like them. I also put everything into `agents.rst` and added an extra example. While going over the 372's usage I found something that doesn't seem right to me (https://github.com/simonsobs/socs/issues/76), so I changed it a bit in the example.

BrianJKoopman commented on docs/developer/agents.rst:
```diff
@@ -430,3 +430,76 @@ containerized Agents together with the command
 Depending on your host's permissions, this command may need to be run with 
 ``sudo``.
+
+.. _timeout_lock
```
This is missing a colon at the end: It should be:
```
.. _timeout_lock:
```

BrianJKoopman commented on docs/developer/agents.rst:
```diff
@@ -430,3 +430,76 @@ containerized Agents together with the command
+dangerous asynchronous behavior.
+
+The ocs :ref:`TimeoutLock <ocs_twisted_api>` is a lock that implements a few useful additions to the
+standard ``threading.lock``. Mainly:
```
Bullets in .rst aren't indented, this thows an "Unexpected indentation" warning when first compiled. Reformat with a blank line after "Mainly:" and no single space indication on the block.

https://docutils.sourceforge.io/docs/user/rst/quickref.html#bullet-lists

BrianJKoopman commented on docs/developer/agents.rst:
```diff
@@ -430,3 +430,76 @@ containerized Agents together with the command
+        if not acquired:
+            print(f"Lock could not be acquired because it is held by {lock.job}")
+            return False
+        print("Lock acquired!")
```
I think a comment with something like "lock requiring actions here" to show the user where to put their code with respect to the lock, i.e. in the with, not after it, would be good.

BrianJKoopman commented on docs/developer/agents.rst:
```diff
@@ -430,3 +430,76 @@ containerized Agents together with the command
+the operation cannot re-acquire the lock, it will print a message and return
+``False``::
+
+    with self.lock.acquire_timeout(timeout=0, job='acq') as acquired:
```
This example would similarly benefit from where the developer is expect to interject their lock reliant code.

This also references class attributes like `self.lock`, can we adjust the example to establish the two locks before the `with` so that the example stands alone?

jlashner replied:
Was thinking about doing that but using a lock that you initialize right before the `with` statement kind of defeats the purpose of using a lock, so I wouldn't want people to copy and paste that into their code. I could do something like expanding the example so that the lock is initialized globally, and then the acquisition is done within a function.

jlashner commented:
Alright, did another iteration on the `release_and_acquire` example that shows it's usage in the context of a simple agent with an interrupting operation. Hopefully the example makes more sense within this context.
